File: e2e/fixtures/seed.ts
import type { Page } from '@playwright/test';

/**
 * Composable seed helpers for e2e tests.
 *
 * Each helper invokes a Tauri command from the `e2e-seed` feature in the Rust
 * crate (see src-tauri/src/cmd/seed.rs). The binary under test MUST be built
 * with `cargo tauri build --debug --features e2e-seed`.
 *
 * Dates are REQUIRED arguments where they appear. Tests must supply fixed
 * dates rather than relying on a notion of "today" inside the seed layer.
 * Times default to fixed constants (12:00:00, 08:00:00) — times-of-day do not
 * drift and are safe to default.
 */

async function invoke<T>(page: Page, command: string, args: object = {}): Promise<T> {
	return page.evaluate(
		({ command, args }) =>
			(
				window as unknown as {
					__TAURI_INTERNALS__: { invoke: (c: string, a: object) => Promise<unknown> };
				}
			).__TAURI_INTERNALS__.invoke(command, args) as Promise<unknown>,
		{ command, args }
	) as Promise<T>;
}

export interface SeedHelpers {
	user(opts?: { name?: string; avatar?: string }): Promise<unknown>;
	bodyData(opts?: {
		age?: number;
		height?: number;
		weight?: number;
		sex?: string;
		activityLevel?: number;
	}): Promise<unknown>;
	intakeTarget(args: {
		startDate: string;
		endDate: string;
		targetCalories?: number;
		maximumCalories?: number;
	}): Promise<unknown>;
	weightTarget(args: {
		startDate: string;
		endDate: string;
		initialWeight?: number;
		targetWeight?: number;
	}): Promise<unknown>;
	intakeEntry(args: {
		date: string;
		amount: number;
		category?: string;
		time?: string;
		description?: string;
	}): Promise<unknown>;
	weightEntry(args: { date: string; weight: number; time?: string }): Promise<unknown>;

	/** Convenience: minimal profile sufficient to access (app) routes. */
	defaultProfile(args: { startDate: string; endDate: string }): Promise<void>;
}

export function makeSeedHelpers(page: Page): SeedHelpers {
	return {
		user: (opts = {}) => invoke(page, 'seed_user', opts),
		bodyData: (opts = {}) => invoke(page, 'seed_body_data', opts),
		intakeTarget: (args) => invoke(page, 'seed_intake_target', args),
		weightTarget: (args) => invoke(page, 'seed_weight_target', args),
		intakeEntry: (args) => invoke(page, 'seed_intake_entry', args),
		weightEntry: (args) => invoke(page, 'seed_weight_entry', args),

		defaultProfile: async ({ startDate, endDate }) => {
			await invoke(page, 'seed_user', {});
			await invoke(page, 'seed_body_data', {});
			await invoke(page, 'seed_intake_target', { startDate, endDate });
			await invoke(page, 'seed_weight_target', { startDate, endDate });
			await invoke(page, 'seed_weight_entry', { date: startDate, weight: 75.0 });
		}
	};
}

File: src-tauri/src/cmd/seed.rs
//! Composable seed primitives for end-to-end tests.
//!
//! Compiled only when the `e2e-seed` Cargo feature is on. Each command wraps
//! an existing service function with a simplified signature.
//!
//! Dates are deliberately REQUIRED (no defaults). Tests that depend on "today"
//! are flaky across midnight rollovers and run-day variance; the seed layer
//! refuses to inject that risk. Times-of-day are stable constants and may be
//! defaulted.

use tauri::{command, State};

use crate::db::connection::DbPool;
use crate::service::body::BodyData;
use crate::service::intake::{Intake, IntakeTarget, NewIntake, NewIntakeTarget};
use crate::service::user::LibreUser;
use crate::service::weight::{NewWeightTarget, NewWeightTracker, WeightTarget, WeightTracker};

#[command]
pub fn seed_user(
    pool: State<DbPool>,
    name: Option<String>,
    avatar: Option<String>,
) -> Result<LibreUser, String> {
    let name = name.unwrap_or_else(|| "Testuser".to_string());
    let avatar = avatar.unwrap_or_else(|| name.clone());
    crate::service::user::update_user(pool, name, avatar)
}

#[command]
pub fn seed_body_data(
    pool: State<DbPool>,
    age: Option<i32>,
    height: Option<f32>,
    weight: Option<f32>,
    sex: Option<String>,
    activity_level: Option<f32>,
) -> Result<BodyData, String> {
    crate::service::body::update_body_data(
        pool,
        age.unwrap_or(30),
        height.unwrap_or(175.0),
        weight.unwrap_or(75.0),
        sex.unwrap_or_else(|| "MALE".to_string()),
        activity_level.unwrap_or(1.5),
    )
}

#[command]
pub fn seed_intake_target(
    pool: State<DbPool>,
    start_date: String,
    end_date: String,
    target_calories: Option<i32>,
    maximum_calories: Option<i32>,
) -> Result<IntakeTarget, String> {
    let new_target = NewIntakeTarget {
        added: start_date.clone(),
        start_date,
        end_date,
        target_calories: target_calories.unwrap_or(2000),
        maximum_calories: maximum_calories.unwrap_or(2500),
    };
    crate::service::intake::create_intake_target(pool, new_target)
}

#[command]
pub fn seed_weight_target(
    pool: State<DbPool>,
    start_date: String,
    end_date: String,
    initial_weight: Option<f32>,
    target_weight: Option<f32>,
) -> Result<WeightTarget, String> {
    let new_target = NewWeightTarget {
        added: start_date.clone(),
        start_date,
        end_date,
        initial_weight: initial_weight.unwrap_or(75.0),
        target_weight: target_weight.unwrap_or(70.0),
    };
    crate::service::weight::create_weight_target(pool, new_target)
}

#[command]
pub fn seed_intake_entry(
    pool: State<DbPool>,
    date: String,
    amount: i32,
    category: Option<String>,
    time: Option<String>,
    description: Option<String>,
) -> Result<Intake, String> {
    let new_entry = NewIntake {
        amount,
        category: category.unwrap_or_else(|| "s".to_string()),
        added: date,
        time: Some(time.unwrap_or_else(|| "12:00:00".to_string())),
        description,
    };
    crate::service::intake::create_intake(pool, new_entry)
}

#[command]
pub fn seed_weight_entry(
    pool: State<DbPool>,
    date: String,
    weight: f32,
    time: Option<String>,
) -> Result<WeightTracker, String> {
    let new_entry = NewWeightTracker {
        amount: weight,
        added: date,
        time: Some(time.unwrap_or_else(|| "08:00:00".to_string())),
    };
    crate::service::weight::create_weight_tracker_entry(pool, new_entry)
}
